### PR TITLE
Proposal new dsl

### PR DESCRIPTION
**Pull request by stoeffel**
This changes the current dsl to use `-->` in between assertion and expectation.
i.e.
```elm
{-| returns the sum of a list of int

    sum aList -- no expectation, this is not a test

    sum
        [ 41
        , 1
        ]
    --> 42

    sum [1, 2] --> 3
-}
sum : List Int -> Int
sum =
    List.foldl (+) 0
```

This way we end up with nicer docs on package.elm-lang.org.
### basic usage
<details><summary>before</summary>
screen shot 2017-06-05 at 15 52 05

</details>
<details><summary>after (with a oneliner example)</summary>
screen shot 2017-06-05 at 15 53 45

</details>

### multiline assertion
<details><summary>before</summary>
screen shot 2017-06-05 at 15 50 44

</details>
<details><summary>after (with an additional example that doesn't have an expectation)</summary>
screen shot 2017-06-05 at 15 55 00
</details>

### multiline expectation
<details><summary>before</summary>
screen shot 2017-06-05 at 15 50 22

</details>
<details><summary>after</summary>
screen shot 2017-06-05 at 15 49 49

</details>

### Global imports

Imported in top comment of a module
screen shot 2017-06-05 at 15 49 43

### Imports before example

screen shot 2017-06-05 at 15 49 38

### Pros:
* nicer docs (see screenshots)
* oneliners
* allows for examples that don't have an expectation, and it doesn't look weird combined
* `-->` arrow is expressive

### Cons: 
* breaking change
* less familiar for people coming from python or other languages which have doc-tests.

### CODE REVIEW

**Comment by stoeffel** [comment]
This is not final and I would love some feedback.
Pinging some people who might care. Sorry if I missed someone 🙇 

@eeue56 @zwilias @cuducos @rtfeldman @ohanhi

**Comment by eeue56** [comment]
Visually, this looks worse to me, with the exception of the "output" line. Couldn't we keep the existing `>>>` and `...`, _and_ use `-->` to denote the expectation?

**Comment by zwilias** [comment]
- doc-test-only imports are great; but could also be supported with the current DSL
- I like the visual appearance in the generated docs - emphasis on the code, the expected output as a comment.
- copy-paste-able code in examples seems like a good idea to me, and this enables that.
- not completely sold on `--.` - the vertical drop of the dot is a little jarring and draws my attention.

**Comment by rtfeldman** [comment]
I'm with @zwilias on this proposal...I really like it!

In particular I'm a big fan of how this reads like normal code examples in documentation. I think this example would look beautiful to me even if I did not know doctests existed:

```elm
 {-| Reverses a list.
  		  
    List.reverse [ "a", "b", "c" ]
    --> [ "c", "b", "a" ]

  -}
```

Compare to the current equivalent:

```elm
 {-| Reverses a list.
  		  
    >>> List.reverse [ "a", "b", "c" ]
    [ "c", "b", "a" ]

  -}
```

I could figure it out, sure, but the whole time I'd be wondering "why is the `>>>` there?"

**Comment by rtfeldman** [comment]
@stoeffel you may have addressed this before, but would it be possible to do the imports without the `--|` at all?

For example:

```elm
{-| Does some funky stuff similar to `Dict.fromList`.

    import Dict exposing (Dict, fromList)


    foo 1 "bar"
    --> fromList [(1, "bar")]

-}
```

Then at the top of the file:

```
# Foo

(Description of Foo module goes here.)

Code examples on this page assume the following imports:

    import Dict exposing (Dict, fromList)
    import String exposing (join)
```

I base this on @jessitron's comment some time ago that "code examples should include imports" so you can copy/paste them into an editor to play around with them, and have them Just Work.

**Comment by rtfeldman** [comment]
> not completely sold on --. - the vertical drop of the dot is a little jarring and draws my attention.

I agree. I'd be curious what using multiple `-->` would look like, although I realize that's more parsing work. 😄

**Comment by zwilias** [comment]
## `--.`

```elm
rev
    [ 41
    , 1
    ]
--> [ 1
--. , 41
--. ]
```

## `-->`
```elm
rev
    [ 41
    , 1
    ]
--> [ 1
--> , 41
--> ]
```

## `---`
```elm
rev
    [ 41
    , 1
    ]
--> [ 1
--- , 41
--- ]
```

**Comment by eeue56** [comment]
I think that these examples should _not_ be considered without looking at how they look in the whole page. Doc tests are not code samples. They are small, tiny examples to help people understand how the function is used, and what to expect back from it. So to highlight this change, each example needs to go through the doc-preview with the full context of the full documentation. Doc tests generally should _not_ replace a full, copy-pastable example.

**Comment by stoeffel** [comment]
> I think that these examples should not be considered without looking at how they look in the whole page. Doc tests are not code samples. They are small, tiny examples to help people understand how the function is used, and what to expect back from it. So to highlight this change, each example needs to go through the doc-preview with the full context of the full documentation. 

Totally agree, screenshots are from doc-preview. The main goal of doc-tests is to create nice verified examples for the user. IMO this new api is nicer to read and feels more familiar for a user who is not familiar to doc-tests.

> Doc tests generally should not replace a full, copy-pastable example.

Agreed!

**Comment by stoeffel** [comment]
I like `---` best, from the three options from @zwilias.

**Comment by cuducos** [comment]
Hum… am I the only one to disagree? I tend to like the `>>>`/`...` although I admit the new syntax gives us beautiful documentation. My main concern regarding this new DSL is about how to differentiate comment from code:


```elm
{-| This is a very complex function that requires a quite long
    comment block. Surely the comment block would be more
    readable if we could use a 2nd paragraph.

    Now I'm not sure if this 2nd paragraph is human readable
    text or Elm code for elm-doc-test.

    computeTheAnswerTo life universe everything
    --> 42
-}
```

I think that the `>>>` would be a bit more expressive in this case:

```elm
{-| This is a very complex function that requires a quite long
    comment block. Surely the comment block would be more
    readable if we could use a 2nd paragraph.

    Now I'm pretty sure this 2nd paragraph is human readable
    text, not Elm code for elm-doc-test.

    >>> computeTheAnswerTo life universe everything
    42
-}
```

In spite of that I think @eeue56 suggestion is expressive too:

```elm
{-| This is a very complex function that requires a quite long
    comment block. Surely the comment block would be more
    readable if we could use a 2nd paragraph.

    Now I'm pretty sure this 2nd paragraph is human readable
    text, not Elm code for elm-doc-test.

    >>> computeTheAnswerTo life universe everything
    --> 42
-}
```

**Comment by rtfeldman** [comment]
@cuducos The Markdown parser always treats the second paragraph as code (because it's indented), so this doesn't work regardless of whether you're writing doctests. 🙂

screen shot 2017-06-06 at 7 02 36 am

**Comment by stoeffel** [comment]
also the latest version of elm-format turns that into:

```elm

{-| This is a very complex function that requires a quite long
comment block. Surely the comment block would be more
readable if we could use a 2nd paragraph.

    Now I'm not sure if this 2nd paragraph is human readable

text or Elm code for elm-doc-test.

    computeTheAnswerTo life universe everything
    --> 42

-}
```

**Comment by stoeffel** [comment]
My main concerns with the current implementation are:
1. beginner might be confused by `>>>`.
2. docs with verified and unverified examples would look weird.
Where this looks natural.
```elm
{-| This is my cool function.

    -- basic usage in your view function
    view : Html msg
    view =
        case coolFunc 42 of
            Cool -> showSomeCoolStuff
            NotCool -> text "uncool, man"

    -- Cool values

    coolFunc 42    --> Cool

    coolFunc 10000 --> Cool

    -- Uncool values

    coolFunc 1 --> Uncool

-}
```
screen shot 2017-06-06 at 08 50 07

**Comment by stoeffel** [comment]
fyi: I will open a new issue for discussing `import`s. This would just change it to be `--|` instead of `>>>` the functionality stays the same. But I have some other ideas around improving `import`.

**Comment by jwoudenberg** [comment]
I really like the proposed change. Also agree with `---` looking nicer than `--.`.

While I agree doc tests are not examples, it's generally nice if examples are doc tests to ensure examples actually work. This change makes that very easy.

With regards to imports: I like the idea about them just being normal `import` statements in an example, but also understand the worry about that creating a lot of visual noise. I wonder if it makes sense to restrict global import to not include `as` or `exposing` clauses, to ensure each doc test is readable on its own.

**Comment by stoeffel** [comment]
> While I agree doc tests are not examples, it's generally nice if examples are doc tests to ensure examples actually work. This change makes that very easy.

With that in mind, I'm also thinking about renaming `elm-doc-test` to something like `elm-verified-docs` or `elm-verified-examples`. Will raise a separate issue for that discussion though.

**Comment by zwilias** [comment]
> I wonder if it makes sense to restrict global import to not include as or exposing clauses, to ensure each doc test is readable on its own.

For modules like `Json.Decode.Extra` is makes sense to assume an `import Json.Decode exposing (..)` - and I'm pretty sure readability of those docs would suffer if that weren't possible...

If anyone's interested, I tried out the proposed DSL on the `Json.Decode.Extra` docs.

- docs.json before and after: https://gist.github.com/zwilias/860a0f19ee69f23ec4243b549437ed85
- upload here: http://package.elm-lang.org/help/docs-preview
- I especially recommend looking at `optionalField` which improved a lot, imho

**Comment by eeue56** [comment]
Given the example of `Cool`, it looks in that highlighting as if the package creator just thinks some of the functions are cool - it doesn't stand out to me that it is the expected output, since it is highlighted as a comment, not code

**Comment by stoeffel** [comment]
hmmm... maybe that was a bad example. will look at the real-world example of `Json.Decode.Extra` in a bit. I think that I saw `-->` in docs already, so I think users might be used to that notation.

**Comment by cuducos** [comment]
> The Markdown parser always treats the second paragraph as code (because it's indented), so this doesn't work regardless of whether you're writing doctests.

Thanks @rtfeldman — my mental compiler was buggy and I haven't considered that. So… well… probably @stoeffel know better than me if the `>>>` is really a problem for beginners. I still thing it's expressive in reinforcing that the line is code… **but** as I come from a Python background this is totally biased, I admit it.

So let's embrace the new DSL and – hell yeah – one liners ; ) I know this might sound ironic but it's not. New syntax is expressive for expectations, beautiful in docs, and way better for imports (even if that is discussed in the new topic).

**Comment by stoeffel** [comment]
@zwilias 
> If anyone's interested, I tried out the proposed DSL on the Json.Decode.Extra docs.

I think the new DSL makes the docs for Json.Decode.Extra way better.

**Comment by eeue56** [comment]
Can you post a screenshot? Having network errors uploading here

**Comment by ohanhi** [comment]
Here's a small portion of the docs with the proposed DSL:

screenshot from 2017-06-06 19-25-21

**Comment by avh4** [comment]
Why not make the special assertion syntax be `|> Expect.equal`:

```elm
{-| returns the sum of a list of int

    sum aList -- no expectation, this is not a test

    sum
        [ 41
        , 1
        ]
    |> Expect.equal 42

    sum [1, 2] |> Expect.equal 3
-}
sum : List Int -> Int
sum =
    List.foldl (+) 0
```

**Comment by eeue56** [comment]
@avh4, I have thought the same previously, but have preferred the "repl" feel of the examples (since you don't confuse things with adding `Expect` into the examples)

Consider the novice - they may copy everything, including the `expect`, without realising it's a testing thing

**Comment by stoeffel** [comment]
I personally don't like that. I want it to become verified examples (a tool
for writing better docs) and not a replacement for tests.

**Comment by avh4** [comment]
What about `==`, then?  And the verification can just be than any top-level expressions that are Bools should eval to True?

**Comment by eeue56** [comment]
If expressions have to result in bools, then you lose the really nice error messages you get from diffing the objects (e.g in elm-html-test)

**Comment by stoeffel** [comment]
I removed `--|` from imports. I still will start a new discussion about imports once this is merged.

**Comment by stoeffel** [comment]
sounds like everyone is onboard with this. discussed it with a few people at elm-europe. Would be cool though if someone could make a quick review.

File: example/Mock.elm [hunk 1]
```diff
@@ -1,45 +1,94 @@
 module Mock exposing (..)
 
-{-| imports for examples
-    >>> import Dict exposing (Dict, fromList)
-    >>> import String exposing (join)
+{-|
+
+@docs add, sum, bar, rev, IsItCool, coolFunc
+
+    -- Imports for doc-tests
```
Comment (eeue56) [inline example/Mock.elm]: This would be more obvious as English (and not code) if it said "These imports are only used in doc-tests"

Comment (stoeffel) [inline example/Mock.elm]: SGTM

File: src/DocTest/Parser.elm [hunk 2]
```diff
@@ -11,15 +11,31 @@ parse str =
...
+            |> List.map filterNotDocTest
+            |> List.filterMap toTest
+    }
+
+
+splitOneLiners : String -> List String
+splitOneLiners str =
+    str
+        |> Regex.replace Regex.All
+            (Regex.regex "\\s\\-\\->\\s")
```
Comment (eeue56) [inline src/DocTest/Parser.elm]: lazy-loading runtime error! needs to go to top level

File: src/DocTest/Parser.elm [hunk 3]
```diff
@@ -11,15 +11,31 @@ parse str =
...
+            (\_ -> "\n    --> ")
```
Comment (eeue56) [inline src/DocTest/Parser.elm]: this function seems strange enough that a name would really help it make sense, IMO

**Pull request merged**